core/src/wallet/ripple/api_impl/RippleLikeTransactionApi.cpp

```diff
@@ -43,6 +43,29 @@
 
 namespace ledger {
     namespace core {
+        // a helper to write VLE fields
+        bool writeLengthPrefix(BytesWriter& writer, uint64_t size) {
+            // encoding here: https://developers.ripple.com/serialization.html#length-prefixing
+            if (size <= 192) {
+                writer.writeByte(size & 0xFF);
+
+                return true;
+            } else if (size <= 12480) {
+                uint16_t x = 193 + ((((size >> 8) & 0x00FF) - 193) << 8) + (size & 0x00FF);
```

Comment:
for me this code do something different from the reference one here https://github.com/ripple/rippled/blob/fa57859477441b60914e6239382c6fba286a0c26/src/ripple/protocol/impl/Serializer.cpp#L313
For example for size = 193
ref:
Byte 0 = 193
Byte 1 = 0
ours:
Byte 0 = 64
Byte 1 = 130

Reply:
That code of theirs seems broken. Since they do a `length -= 193`, the second byte AND mask is wrong compared to the official formula they provide in the official documentation.

>  If the field contains 193 to 12480 bytes of data, the first two bytes indicate the length of the field with the following formula:
> 
> 193 + ((byte1 - 193) * 256) + byte2

So if `length` is two bytes long, `length & 0xFF00` is the first byte padded by 8 bits to the left, to show up, and `length & 0x00FF` is the second one. In order to get the first byte, since it’s padded, we need to move it on 8 bits by moving it 8-bit right. Hence:

```
193 + ((byte1 - 193) * 256) + byte2
193 + ((((length >> 8) & 0x00FF) - 193) << 8) + (length & 0x00FF)
```

You end up with my exact formula. I don’t know why they do it this way, though. Something I’m interested in is: do you see how to change the unit tests to see whether the implementation from this PR is wrong? If I did it wrong, the unit tests should show it :P

Reply:
I think in the formula from documentation `193 + ((byte1 - 193) * 256) + byte2` `byte1` and `byte2` is the resulting bytes, not the bytes of `length`s and with this understanding they code is correct.

Reply:
Yeah actually that’s weird, it looks like which “byte1” or “byte2” comes first is ambiguous. Obviously, if `193` has defined as `byte1 << 8 + byte2`, `byte1` is `0`, so `(byte1 - 193)` is a mess. I need to check further, then.

Reply:
(The interesting part is that they do signed arthimethic while I do unsigned one, which might be why there’s a difference; I might switch to their implementation because when I switch my implementation to signed arithmetic (basically, replace the & 0xFF with * 256), I end up with the same result).